core/broadcast.go

```diff
@@ -0,0 +1,308 @@
+
+// set is a simple interface to keep tracks of all the packet hashes that we
+// have rebroadcast already
+// TODO: check if having a map makes more sense.
+type set interface {
+	put(hash)
+	exists(hash) bool
+}
+
+type arraySet struct {
+	hashes [][]byte
+}
+
+func (a *arraySet) put(hash hash) {
+	for _, h := range a.hashes {
+		if bytes.Equal(h, hash) {
+			return
+		}
+	}
+	a.hashes = append(a.hashes, hash)
+}
+
+func (a *arraySet) exists(hash hash) bool {
+	for _, h := range a.hashes {
+		if bytes.Equal(h, hash) {
+			return true
+		}
+	}
+	return false
+}
+
+type broadcastPacket = *drand.DKGPacket
+
+// dispatcher maintains a list of worker assigned one destination and pushes the
+// message to send to the right worker
+type dispatcher struct {
+	sync.Mutex
+	senders []*sender
+	stopped bool
+}
+
+func newDispatcher(l log.Logger, client net.ProtocolClient, to []*key.Node, us string) *dispatcher {
+	var senders = make([]*sender, 0, len(to)-1)
+	for _, node := range to {
+		if node.Address() == us {
+			continue
+		}
+		sender := newSender(l, client, node)
+		go sender.run()
+		senders = append(senders, sender)
+	}
+	return &dispatcher{
+		senders: senders,
+	}
+}
+
+func (d *dispatcher) broadcast(p broadcastPacket) {
+	if d.isStopped() {
+		return
+	}
+	for _, i := range rand.Perm(len(d.senders)) {
+		d.senders[i].sendPacket(p)
+	}
+}
+
+func (d *dispatcher) stop() {
+	for _, sender := range d.senders {
+		sender.stop()
+	}
+	d.Lock()
+	defer d.Unlock()
+	d.stopped = true
+}
+
+func (d *dispatcher) isStopped() bool {
+	d.Lock()
+	defer d.Unlock()
+	return d.stopped
+}
+
+// size of the receiving queue for a sender channel
+// XXX should it be dynamic ? Reason to fix it is that after some pushed packets
+// if the connection still haven't sent them, that means there's probably
+// something wrong with the connection and we shouldn't push more.
+const senderQueueSize = 10
+
+type sender struct {
+	l       log.Logger
+	client  net.ProtocolClient
+	to      net.Peer
+	newCh   chan broadcastPacket
+	closeCh chan bool
+}
+
+func newSender(l log.Logger, client net.ProtocolClient, to net.Peer) *sender {
+	return &sender{
+		l:       l,
+		client:  client,
+		to:      to,
+		newCh:   make(chan broadcastPacket, senderQueueSize),
+		closeCh: make(chan bool, 1),
+	}
+}
+
+func (s *sender) sendPacket(p broadcastPacket) {
+	select {
+	case s.newCh <- p:
+	default:
+		s.l.Debug("broadcast", "sender queue full", "endpoint", s.to.Address())
+	}
+}
+
+func (s *sender) run() {
+	// indicator to drain the channel until the end before stopping
+	var toFinish = -1
+	for {
+		select {
+		case newPacket := <-s.newCh:
+			err := s.client.BroadcastDKG(context.Background(), s.to, newPacket)
+			if err != nil {
+				s.l.Debug("broadcast", "sending out", "error to", s.to.Address(), "err:", err)
+			} else {
+				s.l.Debug("broadcast", "sending out", "to", s.to.Address())
+			}
+			if toFinish == 0 {
+				return
+			}
+		case <-s.closeCh:
+			toFinish = len(s.newCh)
```

Comment:
rather than an explicit `s.closeCh`, can we have `close()` perform a `close(s.newCh)` as a way of preventing any new items from being stored there. then in the `run` thread, you can just always drain the channel:
`for item := range s.newCh {...}` which will end when it finishes draining and reaches a closed indicator for the channel.